SU2_CFD/src/numerics/turbulent/turb_sources.cpp

```diff
@@ -76,6 +76,179 @@ CNumerics::ResidualType<> CSourcePieceWise_TurbSA::ComputeResidual(const CConfig
 //  AD::SetPreaccIn(TurbVar_Grad_i[0], nDim);
 //  AD::SetPreaccIn(Volume); AD::SetPreaccIn(dist_i);
 
+//  BC Transition Model variables
+  su2double vmag, rey, re_theta, re_theta_t, re_v;
+  su2double tu , nu_cr, nu_t, nu_BC, chi_1, chi_2, term1, term2, term_exponential;
+
+  // Set the boolean here depending on whether the point is closest to a rough wall or not.
+  roughwall = (roughness_i > 0.0);
+
+  if (incompressible) {
+    Density_i = V_i[nDim+2];
+    Laminar_Viscosity_i = V_i[nDim+4];
+  }
+  else {
+    Density_i = V_i[nDim+2];
+    Laminar_Viscosity_i = V_i[nDim+5];
+  }
+
+  Residual        = 0.0;
+  Production      = 0.0;
+  Destruction     = 0.0;
+  CrossProduction = 0.0;
+  Jacobian_i[0]   = 0.0;
+
+  gamma_BC = 0.0;
+  vmag = 0.0;
+  tu   = config->GetTurbulenceIntensity_FreeStream();
+  rey  = config->GetReynolds();
+
+  if (nDim==2) {
+    vmag = sqrt(V_i[1]*V_i[1]+V_i[2]*V_i[2]);
+  }
+  else {
+    vmag = sqrt(V_i[1]*V_i[1]+V_i[2]*V_i[2]+V_i[3]*V_i[3]);
+  }
+
+  /*--- Evaluate Omega ---*/
+
+  Omega = sqrt(Vorticity_i[0]*Vorticity_i[0] + Vorticity_i[1]*Vorticity_i[1] + Vorticity_i[2]*Vorticity_i[2]);
+
+  /*--- Rotational correction term ---*/
+
+  if (rotating_frame) { Omega += 2.0*min(0.0, StrainMag_i-Omega); }
+
+  if (dist_i > 1e-10) {
+
+    /*--- Production term ---*/
+
+    dist_i_2 = dist_i*dist_i;
+    nu = Laminar_Viscosity_i/Density_i;
+
+    /*--- Modified values for roughness ---*/
+    /*--- Ref: Aupoix, B. and Spalart, P. R., "Extensions of the Spalart-Allmaras Turbulence Model to Account for Wall Roughness,"
+     * International Journal of Heat and Fluid Flow, Vol. 24, 2003, pp. 454-462. ---*/
+    /* --- See https://turbmodels.larc.nasa.gov/spalart.html#sarough for detailed explanation. ---*/
+
+    Ji = TurbVar_i[0]/nu  + cr1*(roughness_i/(dist_i+EPS)); //roughness_i = 0 for smooth walls and Ji remains the same, changes only if roughness is specified.
+    Ji_2 = Ji*Ji;
+    Ji_3 = Ji_2*Ji;
+    fv1 = Ji_3/(Ji_3+cv1_3);
+
+    /*--- Using a modified relation so as to not change the Shat that depends on fv2. ---*/
+    fv2 = 1.0 - TurbVar_i[0]/(nu+TurbVar_i[0]*fv1);   // From NASA turb modeling resource and 2003 paper
+
+    ft2 = ct3*exp(-ct4*Ji_2);
+    S = Omega;
+    inv_k2_d2 = 1.0/(k2*dist_i_2);
+
+    Shat = S + TurbVar_i[0]*fv2*inv_k2_d2;
+    Shat = max(Shat, 1.0e-10);
+    inv_Shat = 1.0/Shat;
+
+//    Original SA model
+//    Production = cb1*(1.0-ft2)*Shat*TurbVar_i[0]*Volume;
+
+    if (transition) {
```

Comment:
A small comment. 
It looks like the BC transition model is only valid with the no-ft2 version (https://turbmodels.larc.nasa.gov/sa-bc_1eqn.html).

Reply:
Thanks for pointing it out.
"the ft2 term present in (SA) may cause an undesirable delay in transition to turbulence in the RANS region, and should be avoided".
I will take it into account.